TombEngine/Objects/TR3/Entity/tr3_civvy.cpp

```diff
@@ -2,418 +2,594 @@
 		}
 		else
 		{
-			if (item->AIBits)
-				GetAITarget(creature);
+			if (item.AIBits)
+				GetAITarget(&creature);
 			else
-				creature->Enemy = LaraItem;
+				creature.Enemy = CivvyFindNearTarget(itemNumber, CivvyExcludedTargets, CIVVY_AWARE_RANGE);
 
 			AI_INFO AI;
-			CreatureAIInfo(item, &AI);
+			CreatureAIInfo(&item, &AI);
 
-			AI_INFO laraAI;
-			if (creature->Enemy == LaraItem)
+			float distance2D = 0;
+			int angle = 0;
+			if (creature.Enemy->IsLara())
 			{
-				laraAI.angle = AI.angle;
-				laraAI.distance = AI.distance;
+				targetAngle = AI.angle;
+				targetDistance = AI.distance;
 			}
 			else
 			{
-				int laraDz = LaraItem->Pose.Position.z - item->Pose.Position.z;
-				int laraDx = LaraItem->Pose.Position.x - item->Pose.Position.x;
-				laraAI.angle = phd_atan(laraDz, laraDx) - item->Pose.Orientation.y;
-				laraAI.distance = SQUARE(laraDx) + SQUARE(laraDz);
+				int targetDx = creature.Enemy->Pose.Position.x - item.Pose.Position.x;
+				int targetDz = creature.Enemy->Pose.Position.z - item.Pose.Position.z;
+				targetAngle = phd_atan(targetDz, targetDx) - item.Pose.Orientation.y;
+				targetDistance = SQUARE(targetDx) + SQUARE(targetDz);
 			}
 
-			GetCreatureMood(item, &AI, true);
+			//If Lara was placed by system (CreatureAIInfo), not because she were a real target. Then delete the target.
+			if (creature.Enemy->IsLara() && !creature.HurtByLara)
+				creature.Enemy = nullptr;
+
+			GetCreatureMood(&item, &AI, true);
 
-			if (creature->Enemy == LaraItem &&
+			if (creature.Enemy == LaraItem &&
 				AI.distance > CIVVY_ESCAPE_RANGE &&
 				AI.enemyFacing < ANGLE(67.0f) &&
 				AI.enemyFacing > -ANGLE(67.0f))
 			{
-				creature->Mood = MoodType::Escape;
+				creature.Mood = MoodType::Escape;
 			}
 
-			CreatureMood(item, &AI, true);
+			CreatureMood(&item, &AI, true);
 
-			angle = CreatureTurn(item, creature->MaxTurn);
+			angle = CreatureTurn(&item, creature.MaxTurn);
 
-			auto* realEnemy = creature->Enemy;
-			creature->Enemy = LaraItem;
 
-			if ((laraAI.distance < CIVVY_AWARE_RANGE || item->HitStatus || TargetVisible(item, &laraAI)) &&
-				!(item->AIBits & FOLLOW))
+			//TODO: Make a way to detect entities if it has an AI Object (A current issue for other entities too).
+			if ( item.HitStatus ||
+				 (targetDistance <= CIVVY_AWARE_RANGE && 
+				  creature.Enemy != nullptr && TargetVisible(&item, &AI) &&
+				  !(item.AIBits & FOLLOW))
+				)
 			{
-				if (!creature->Alerted)
-					SoundEffect(SFX_TR3_AMERCAN_HOY, &item->Pose);
+				if (!creature.Alerted)
+					SoundEffect(SFX_TR3_AMERCAN_HOY, &item.Pose);
 
 				AlertAllGuards(itemNumber);
 			}
-			creature->Enemy = realEnemy;
 
-			switch (item->Animation.ActiveState)
+
+			switch (item.Animation.ActiveState)
 			{
-			case CIVVY_WAIT:
-				if (creature->Alerted || item->Animation.TargetState == CIVVY_STATE_RUN_FORWARD)
-				{
-					item->Animation.TargetState = CIVVY_STATE_IDLE;
-					break;
-				}
+				case CIVVY_STATE_IDLE:
```

Comment:
These case blocks appear to be tabbed in. Coding conventions suggest avoiding this.

Reply:
Argh, yes.

Well, I can adapt to it.

Reply:
Done

Reply:
The odd switch case syntax is a relic of C. They look out of place, I know. :P